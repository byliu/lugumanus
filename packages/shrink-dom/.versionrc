{
  "types": [
    {
      "type": "feat",
      "section": "✨ 新功能"
    },
    {
      "type": "fix",
      "section": "🐛 Bug 修复"
    },
    {
      "type": "docs",
      "section": "📝 文档"
    },
    {
      "type": "style",
      "section": "💄 样式"
    },
    {
      "type": "refactor",
      "section": "♻️ 重构"
    },
    {
      "type": "perf",
      "section": "⚡️ 性能优化"
    },
    {
      "type": "test",
      "section": "✅ 测试"
    },
    {
      "type": "chore",
      "section": "🔧 构建/工具"
    },
    {
      "type": "revert",
      "section": "⏪ 回退"
    },
    {
      "type": "build",
      "section": "📦 打包"
    },
    {
      "type": "ci",
      "section": "👷 CI相关"
    }
  ]
}
