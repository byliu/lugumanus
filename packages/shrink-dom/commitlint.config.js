export default {
  extends: ['@commitlint/config-conventional'],
  rules: {
    'type-enum': [
      2,
      'always',
      [
        'feat', // 新功能
        'fix', // 修复bug
        'docs', // 文档更新
        'style', // 代码风格修改（不影响代码运行）
        'refactor', // 重构
        'perf', // 性能优化
        'test', // 测试
        'chore', // 构建过程或辅助工具的变动
        'revert', // 回退
        'build', // 打包
        'ci', // CI相关变更
      ],
    ],
    'subject-case': [0], // 这可以放宽主题的大小写限制
  },
};
