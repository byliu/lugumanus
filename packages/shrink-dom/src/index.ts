export { DOMDomainAnalyzer } from './dom-domain-analyzer';
export { DOMShrinker } from './shrink-dom';
export { DOMContentExtractor } from './dom-content-extractor';

export type {
  ExtractorOptions,
  JsonNode,
  OptimizedTemplate,
  PossibleTemplate,
  ChosenTemplate,
} from './types';
