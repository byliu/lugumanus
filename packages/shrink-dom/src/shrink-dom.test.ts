import { describe, expect, test } from 'bun:test';
import path from 'node:path';
import { JSDOM } from 'jsdom';
import { DOMContentExtractor } from './dom-content-extractor';
import { DOMDomainAnalyzer } from './dom-domain-analyzer';
import { DOMShrinker } from './shrink-dom';

describe('DOMShrinker', () => {
  // 测试基本压缩功能
  test('应该正确压缩简单HTML', () => {
    const html = '<div><p>Hello</p><p>World</p></div>';
    const result = DOMShrinker.compressHTMLString(html);

    // 确保结果包含模板定义和引用
    expect(result).toContain('T1:');
    expect(result).toContain('<p>Hello</p>');
  });

  // 测试带有属性的HTML压缩
  test('应该正确处理带有属性的HTML', () => {
    const html = `
      <div class="container">
        <button class="btn" type="button">Click me</button>
        <button class="btn" type="submit">Submit</button>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(html);

    // 确保结果包含按钮模板
    expect(result).toContain('btn');
    expect(result).toContain('button');
  });

  // 测试嵌套结构的HTML压缩
  test('应该正确处理嵌套结构', () => {
    const html = `
      <div>
        <ul>
          <li><span>Item 1</span></li>
          <li><span>Item 2</span></li>
          <li><span>Item 3</span></li>
        </ul>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(html);

    // 确保结果包含列表项模板
    expect(result).toContain('<li>');
    expect(result).toContain('<span>');
  });

  // 测试自定义配置选项
  test('应该尊重自定义配置选项', () => {
    const html = `
      <div>
        <p>Paragraph 1</p>
        <p>Paragraph 2</p>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(html, {
      templateIdPrefix: 'Template',
      minTemplateOccurrences: 2,
    });

    // 确保使用了自定义前缀
    expect(result).toContain('Template1:');
  });

  // 测试复杂的真实世界HTML
  test('应该处理复杂的真实世界HTML', () => {
    const html = `
      <div class="card-container">
        <div class="card">
          <div class="card-header">
            <h3>Product 1</h3>
            <span class="price">$19.99</span>
          </div>
          <div class="card-body">
            <p>This is product 1 description</p>
            <button class="btn">Add to Cart</button>
          </div>
        </div>
        <div class="card">
          <div class="card-header">
            <h3>Product 2</h3>
            <span class="price">$29.99</span>
          </div>
          <div class="card-body">
            <p>This is product 2 description</p>
            <button class="btn">Add to Cart</button>
          </div>
        </div>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(html);

    // 确保识别出卡片模板
    expect(result).toContain('card');
    expect(result).toContain('card-header');
    expect(result).toContain('card-body');
  });

  // 测试空HTML处理
  test('应该优雅地处理空HTML', () => {
    const html = '<div></div>';
    const result = DOMShrinker.compressHTMLString(html);

    // 应该返回简单的div，没有模板
    expect(result).toContain('<div></div>');
  });

  // 测试直接使用DOM元素
  test('应该支持直接使用DOM元素', () => {
    const dom = new JSDOM('<div><p>Test</p><p>Test</p></div>');
    const shrinker = new DOMShrinker();
    const result = shrinker.compressHTML(dom.window.document.body);

    // 确保结果包含模板
    expect(result).toContain('T1:');
    expect(result).toContain('<p>Test</p>');
  });

  // 测试不同的模板深度设置
  test('应该尊重最小模板深度设置', () => {
    const html = `
      <div>
        <p><span>Text 1</span></p>
        <p><span>Text 2</span></p>
      </div>
    `;

    // 使用较高的最小深度
    const result1 = DOMShrinker.compressHTMLString(html, {
      minTemplateDepth: 3,
    });

    // 使用较低的最小深度
    const result2 = DOMShrinker.compressHTMLString(html, {
      minTemplateDepth: 1,
    });

    // 结果应该不同
    expect(result1).not.toEqual(result2);
  });

  // 测试带有注释的HTML
  test('应该正确处理带有注释的HTML', () => {
    const html = `
      <div>
        <!-- 这是一个注释 -->
        <p>Paragraph 1</p>
        <p>Paragraph 2</p>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(html);

    // 确保结果包含段落模板但忽略注释
    expect(result).toContain('<p>');
    expect(result).not.toContain('这是一个注释');
  });

  // 测试带有特殊字符的HTML
  test('应该正确处理带有特殊字符的HTML', () => {
    const html = `
      <div>
        <p>特殊字符: &lt; &gt; &amp; &quot; &apos;</p>
        <p>特殊字符: &lt; &gt; &amp; &quot; &apos;</p>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(html);

    // 确保特殊字符被正确处理
    expect(result).toContain('特殊字符');
    expect(result).toContain('&lt;');
    expect(result).toContain('&gt;');
    expect(result).toContain('&amp;');
  });

  // 使用快照测试验证压缩结果的一致性
  test('应该生成一致的压缩输出并匹配快照', () => {
    const complexHtml = `
      <div class="article">
        <header>
          <h1>文章标题</h1>
          <div class="meta">
            <span class="author">作者名</span>
            <span class="date">2023-04-01</span>
          </div>
        </header>
        <section class="content">
          <p>第一段落内容</p>
          <p>第二段落内容</p>
          <ul class="points">
            <li>要点一</li>
            <li id="point-2" class="point">要点二</li>
            <li>要点三</li>
            <li>小节奏</li>
          </ul>
        </section>
        <footer>
          <div class="tags">
            <span class="tag2">标签1</span>
            <span class="tag">标签2</span>
            <span class="tag">标签3</span>
          </div>
          <button class="like">点赞</button>
          <button class="share">分享</button>
          <div class="link">
            <a href="https://www.xiaolugu.com">百度</a>
            <a href="https://www.baidu.com">百度</a>
          </div>
        </footer>
      </div>
    `;

    const result = DOMShrinker.compressHTMLString(complexHtml, {
      minTemplateOccurrences: 2,
      minTemplateDepth: 2,
    });

    // 使用快照测试确保压缩结果的一致性
    expect(result).toMatchSnapshot();
  });

  test('处理真实的DOM结构', async () => {
    const html = await Bun.file(path.join(import.meta.dir, 'mock/mock.html')).text();
    const dom = new JSDOM(html);
    globalThis.window = dom.window as unknown as Window & typeof globalThis;
    globalThis.document = dom.window.document;

    const analyzer = new DOMDomainAnalyzer();
    const result = analyzer.getAnalyzedHTML(dom.window.document.documentElement);
    const analyzedDOM = new JSDOM(result);

    const extractor = new DOMContentExtractor();
    const result2 = extractor.extract(analyzedDOM.window.document.documentElement);
    const extractedDOM = new JSDOM((result2 as HTMLElement)?.outerHTML);

    expect(extractedDOM).not.toBeNull();

    const shrinker = new DOMShrinker({
      minTemplateOccurrences: 1,
      minTemplateDepth: 2,
    });
    const result3 = shrinker.compressHTML(extractedDOM.window.document.body);
    expect(result3).toMatchSnapshot();
  });
});
