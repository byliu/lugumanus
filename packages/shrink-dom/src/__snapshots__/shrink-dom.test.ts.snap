// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`DOMShrinker 处理真实的DOM结构 1`] = `
"T1: <span>4.8 分</span>
T2: <div>国际及 中国港澳台</div>
T3: <span>¥ 458 起</span>
T4: <div>￥ 1413.0 起</div>
T5: <div>携程精选<span>2.7 折</span></div>
T6: <span>4.8 /5</span>
T7: <div>温州<span>¥ 230</span></div>
T8: <span><span>4.8 /5</span>超棒</span>
T9: <div><span>上海 南京</span><span>¥ 560 起</span></div>
T10: <span>04-25 去 04-27 回</span>
T11: <div>¥ 688</div>
T12: <div><div><span>上海 南京</span><span>¥ 560 起</span></div><div>携程精选<span>2.7 折</span></div><span>04-25 去 04-27 回</span></div>
T13: <span>24947 人出游</span>
T14: <div>酒店 · 放心住</div>
T15: <div><span>4.8 分</span>|<span>24947 人出游</span></div>
T16: <div>1765条点评<div>¥ 687</div></div>
T17: <div>武汉酒店口碑榜 19℃~33℃</div>
T18: <span>¥ 1558 起</span>
T19: <div>苏州+乌镇+杭州3日2晚跟团游<span>¥ 479 起</span></div>
T20: <div>武汉泛海费尔蒙酒店<div>￥ 1413.0 起</div></div>
T21: <div>武汉洲际酒店<div>￥ 1398.0 起</div></div>
T22: <ul>热门 ABCDEF GHIJ KLMN PQRSTUVW XYZ</ul>
T23: <ul>北京 上海 广州 深圳 成都 杭州 武汉 西安 重庆 青岛 长沙 南京 厦门 昆明 大连 天津 郑州 三亚 济南 福州</ul>
T24: <div><div>苏州+乌镇+杭州3日2晚跟团游<span>¥ 479 起</span></div><div><span>4.9 分</span>|<span>11215 人出游</span></div></div>
T25: <div>价格透明，退票改签不收取 额外费用，保障客人出行</div>
T26: <p type="text">3月27日 (今天)</p>
T27: <div><span><span>4.6 /5</span>超棒</span>¥899</div>
T28: <div>国内<div>国际及 中国港澳台</div></div>
T29: <div><ul>热门 ABCDEF GHIJ KLMN PQRSTUVW XYZ</ul><ul>北京 上海 广州 深圳 成都 杭州 武汉 西安 重庆 青岛 长沙 南京 厦门 昆明 大连 天津 郑州 三亚 济南 福州</ul></div>
T30: <div>支持中文/拼音/英文输入<div>国内<div>国际及 中国港澳台</div></div><div><ul>热门 ABCDEF GHIJ KLMN PQRSTUVW XYZ</ul><ul>北京 上海 广州 深圳 成都 杭州 武汉 西安 重庆 青岛 长沙 南京 厦门 昆明 大连 天津 郑州 三亚 济南 福州</ul></div></div>
T31: <div>武汉万达瑞华酒店<div>￥ 1391.0 起</div></div>
T32: <div>合肥君悦酒店<div>￥ 738.0 起</div></div>
T33: <div>合肥栢景朗廷酒店<div>￥ 708.0 起</div></div>
T34: <div>合肥富力威斯汀酒店<div>￥ 673.0 起</div></div>
T35: <div><div>武汉泛海费尔蒙酒店<div>￥ 1413.0 起</div></div><div>武汉万达瑞华酒店<div>￥ 1391.0 起</div></div><div>武汉洲际酒店<div>￥ 1398.0 起</div></div></div>
T36: <div><div>合肥君悦酒店<div>￥ 738.0 起</div></div><div>合肥栢景朗廷酒店<div>￥ 708.0 起</div></div><div>合肥富力威斯汀酒店<div>￥ 673.0 起</div></div></div>
T37: <div><div>旅游 · 放心玩</div><div>开放数万条旅游产品选择 7*24小时应急救援体系</div></div>

<body><div><div>亲爱的用户,您当前浏览器版本过低,会影响浏览体验 建议升级浏览器<p><a id="14">下载Firefox</a><a id="15">下载Chrome</a></p>或使用手机APP浏览</div><div role="navigation" id="20"><div role="navigation" id="21"><div><div><button aria-label="酒店 按回车键打开菜单" id="31">酒店</button><div><a id="35">国内酒店</a><a id="37">海外酒店</a></div></div><div><button aria-label="机票 按回车键打开菜单" id="40">机票</button><div><a id="44">国内/国际/中国港澳台</a><a id="46">特价机票</a><a id="48">航班动态</a><a id="50">值机选座</a><a id="52">退票改签</a><a id="54">报销凭证</a><a id="56">机场攻略</a><a id="58">定制包机</a></div></div><div><button aria-label="火车票 按回车键打开菜单" id="61">火车票</button><div><a id="65">国内火车票</a><a id="67">国际/中国港澳台</a></div></div><div><button aria-label="旅游 按回车键打开菜单" id="70">旅游</button><div><a id="74">旅游首页</a><a id="76">跟团游</a><a id="78">私家团</a><a id="80">自由行</a><a id="82">邮轮</a><a id="84">一日游</a><a id="86">定制旅游</a><a id="88">周边游</a><a id="90">高端游</a><a id="92">主题游</a><a id="94">游学</a><a id="96">签证</a><a id="98">保险</a><a id="100">企业会奖</a><a id="102">目的地</a></div></div><button aria-label="门票·活动 按回车键打开菜单" id="105">门票·活动</button><div><button aria-label="汽车·船票 按回车键打开菜单" id="110">汽车·船票</button><div><a id="114">汽车票</a><a id="116">船票</a></div></div><div><button aria-label="用车 按回车键打开菜单" id="119">用车</button><div><a id="123">国内租车</a><a id="125">境外租车</a><a id="127">接送机站</a><a id="129">按天包车</a></div></div><button aria-label="攻略·景点 按回车键打开菜单" id="132">攻略·景点</button><button aria-label="旅游地图 按回车键打开菜单" id="137">旅游地图</button><div><button aria-label="全球购 按回车键打开菜单" id="143">全球购</button><div><a id="147">名店购</a><a id="149">银联特惠</a><a id="151">外币兑换</a></div></div><div><button aria-label="礼品卡 按回车键打开菜单" id="154">礼品卡</button><div><a id="158">礼品卡首页</a><a id="160">企业采购</a><a id="162">心意送礼</a><a id="164">礼品卡福袋</a></div></div><button aria-label="携程金融 按回车键打开菜单" id="167">携程金融</button><button aria-label="企业商旅 按回车键打开菜单" id="172">企业商旅</button><button aria-label="老友会 按回车键打开菜单" id="177">老友会</button><button aria-label="关于携程 按回车键打开菜单" id="182">关于携程</button></div></div></div><div><div role="navigation" id="191"><div><div><a id="197">携程旅行网</a><input type="text" placeholder="搜索任何旅游相关" aria-label="搜旅任何旅行相关..." id="203" editable="true" /></div><div><ul><button aria-label="" id="210">登录</button><button id="214">注册</button><button id="219">我的订单</button><button aria-label="携程客服" id="225">联系客服</button><button aria-label="扫码下载携程App或关注微信公众号" id="230"></button></ul><button aria-label="无障碍阅读开关" id="232"></button><button aria-label="关怀版" id="234"></button></div></div></div><div><div><div role="search" id="239"><div><div aria-label="您已进入酒店预订去，该模块包含目的地、入住、离店日期、房间数、住客数、酒店级别以及关键词等输入框。按下tab键可进入输入框。">预订酒店</div><ul><div><label id="251">目的地/酒店名称</label><input type="text" placeholder="城市、机场、区域、地标或酒店名称" aria-label="请通过键盘输入目的地名，例如：上海市。键入关键词时会展示联想浮层，按下TAB键或SHIFT+TAB键即可关闭联想浮层。" value="上海" id="252" editable="true" /></div><div><div><label id="356">入住</label>{T26(0)}</div><div aria-label="">1晚</div><div><label id="362">退房</label>{T26(0)}</div></div><div>房间及住客<p>1 间, 1 位</p></div><div>酒店级别 不限</div><div><label id="424">关键词（选填）</label><input type="text" value="" placeholder="机场/火车站/酒店名称..." aria-label="请输入关键词（选填），键入关键词时会展示联想浮层，按下TAB键或SHIFT+TAB键即可关闭联想浮层。" id="426" editable="true" /></div>搜索</ul></div></div><div role="main" id="464"><div><div><div id="468">酒店 推荐</div><ul><button aria-label="出发地选择上海" id="471">上海</button><button id="472">北京</button><button id="473">广州</button><button id="474">三亚</button><button id="476">更多</button></ul></div><ul><a aria-label="酒店名称上海新天地大世界地铁站亚朵酒店,评分4.8/5,460条点评,价格1148元起" id="485"><div>上海新天地大世界地铁站亚朵酒店{T8(0)}<div>460条点评<div>¥ 1148</div></div></div></a><a aria-label="酒店名称上海虹桥水城路亚朵酒店,评分4.9/5,150条点评,价格688元起" id="502"><div>上海虹桥水城路亚朵酒店{T8(0)}<div>150条点评{T11(0)}</div></div></a><a aria-label="酒店名称如家商旅酒店(上海新国际博览中心张江高科广兰路地铁站店),评分4.6/5,1765条点评,价格687元起" id="519"><div>如家商旅酒店(上海新国际博览中心张江高科广兰路地铁站店){T27(0)}{T16(0,1)}</div></a><a aria-label="酒店名称汉庭酒店(上海虹桥火车站北翟路店),评分4.6/5,1257条点评,价格179元起" id="537"><div>汉庭酒店(上海虹桥火车站北翟路店){T27(0)}{T16(0,1)}</div></a><a aria-label="酒店名称上海虹桥机场城际酒店,评分4.8/5,4106条点评,价格646元起" id="555"><div>上海虹桥机场城际酒店{T8(0)}{T16(0,1)}</div></a><a aria-label="酒店名称上海金桥佳友CitiGO欢阁酒店,评分4.8/5,1365条点评,价格852元起" id="572"><div>上海金桥佳友CitiGO欢阁酒店{T8(0)}{T16(0,1)}</div></a></ul><div><span id="592">当季 热推</span>出发地<div><input aria-label="城市" name="owDCity" type="text" placeholder="城市" value="上海" id="598" editable="true" />{T30()}</div></div><div><div><a id="643">当季热卖·跟团游</a><ul><a aria-label="跟团游名称:普陀山2日1晚跟团游,价格458元起,评分4.8分,24947人已出游" id="648">1<div><div>普陀山2日1晚跟团游{T3(0)}</div>{T15(0,1)}</div></a><a aria-label="跟团游名称:苏州+乌镇+杭州3日2晚跟团游,价格479元起,评分4.9分,11215人已出游" id="662">2{T24(0,1)}</a><a aria-label="跟团游名称:浙江普陀山2日1晚跟团游,价格458元起,评分4.7分,27209人已出游" id="676">3<div><div>浙江普陀山2日1晚跟团游{T3(0)}</div>{T15(0,1)}</div></a><a aria-label="跟团游名称:黄山风景区3日2晚跟团游,价格1558元起,评分4.9分,1906人已出游" id="690">4<div><div>黄山风景区3日2晚跟团游{T18(0)}</div><div>{T1(0)}|{T18(0)}</div></div></a><a aria-label="跟团游名称:苏州+乌镇+杭州3日2晚跟团游,价格479元起,评分4.8分,21950人已出游" id="704">5{T24(0,1)}</a></ul></div><div><a id="721">周末畅游·特价机票</a><ul><a aria-label="上海至南京往返机票,价格560元起,04-25去,04-27回" id="726">1{T12(0,1,2,3)}</a><a aria-label="上海至福州往返机票,价格570元起,04-26去,04-27回" id="743">2{T12(0,1,2,3)}</a><a aria-label="上海至合肥往返机票,价格596元起,03-28去,03-31回" id="760">3{T12(0,1,2,3)}</a><a aria-label="上海至广州往返机票,价格598元起,04-04去,04-05回" id="777">4{T12(0,1,2,3)}</a><a aria-label="上海至南昌往返机票,价格601元起,03-29去,04-01回" id="794">5{T12(0,1,2,3)}</a></ul></div></div></div><div><div><h3>探索 精彩世界</h3><div>出发地<div><input aria-label="城市" name="owDCity" type="text" placeholder="城市" value="上海" id="821" editable="true" />{T30()}</div></div><a id="861">更多目的地</a></div><div><div><div>石家庄{T1(0)}</div>最低价</div>{T7(0,1)}{T7(0,1)}{T7(0,1)}{T7(0,1)}{T7(0,1)}{T7(0,1)}{T7(0,1)}{T5(0,1)}{T7(0,1)}</div></div></div></div><div role="presentation" id="1081"><div><div id="1085">{T17(0)}{T35()}</div><div id="1114">{T17(0)}{T36()}</div><div id="1143">{T17(0)}<div><div>黄山悦榕庄{T4(0)}</div>{T20(0,1)}<div>黄山昱城皇冠假日酒店{T2(0)}</div></div></div><div id="1172"><div>大理酒店口碑榜 6℃~23℃</div><div><div>蜜悦·大理悦湾VILLA半山海景酒店{T2(0)}</div>{T21(0,1)}</div></div><div id="1193">{T17(0)}{T35()}</div><div id="1222">{T17(0)}{T36()}</div></div></div></div><div>企业商旅<div><a id="1265"><div>一站式企业差旅服务 海量差旅产品，全流程服务，智能管控，助力企业成本节省高达30%！</div></a><a id="1270"><div>公对公结算 15分钟极速开通公司账户、30+20天超长账期，自助对账，统一开票、配送！</div></a><a id="1275"><div>企业方案量身定制 强大的产品技术&amp;服务解决方案 ，支持企业规模在500人以上中大型企业定制</div></a></div></div><div><div>携程旅行保障<div><div>{T14(0)}<div>200多个国家121万家酒店 确认订单后不涨价</div></div><div>{T14(0)}<div>近200万条全球航线选择 一站式售后服务体验</div></div><div>{T2(0)}{T25(0)}</div>{T37(0,1)}</div><div><div>{T14(0)}<div>多元化结款产品 合规保障，专项服务</div></div><div>{T14(0)}<div>全城美食套餐性价比高 品质优选，价格保障</div></div>{T37(0,1)}<div>{T14(0)}<div>旅游专项直播随时逛 大BOSS定期开播</div></div></div></div>合作伙伴<div role="navigation" id="1366"><div>携程导航<div><a id="1371">酒店预订</a><a id="1372">海外酒店</a><a id="1373">特价机票</a><a id="1374">国际机票</a><a id="1375">旅游线路</a><a id="1376">景点门票</a><a id="1377">旅游攻略</a><a id="1378">火车票</a><a id="1379">邮轮</a><a id="1380">周末游</a><a id="1381">签证</a><a id="1382">租车</a><a id="1383">欧铁通票</a><a id="1384">自由行</a><a id="1385">会议旅游</a><a id="1386">礼品卡</a><a id="1387">团队游</a><a id="1388">企业商旅</a><button id="1389">查看全部</button></div></div><div>酒店预订<div>携程酒店官网为您提供全球200多个国家和地区、9万多个城市的酒店预订及价格查询服务，覆盖酒店数超过170万家。提供各种住宿类型，包括：酒店、宾馆、旅社、客栈、民宿、经济连锁、酒店公寓、青年旅舍、农家乐、别墅、特色住宿等。同时提供酒店图片、房间照片、酒店电话、酒店地址以及真实用户的酒店点评等各类信息，方便用户选择。<div><a id="1395">北京酒店</a><a id="1396">上海酒店</a><a id="1397">广州酒店</a><a id="1398">深圳酒店</a><a id="1399">南京酒店</a><a id="1400">杭州酒店</a><a id="1401">成都酒店</a><a id="1402">厦门酒店</a><a id="1403">青岛酒店</a><a id="1404">三亚酒店</a><a id="1405">香港酒店</a><a id="1406">澳门酒店</a><a id="1407">西安酒店</a><a id="1408">苏州酒店</a><a id="1409">长沙酒店</a><a id="1410">武汉酒店</a><a id="1411">大连酒店</a><a id="1412">天津酒店</a><a id="1413">重庆酒店</a><a id="1414">长春酒店</a><a id="1415">济南酒店</a><a id="1416">贵阳酒店</a><a id="1417">昆明酒店</a><a id="1418">郑州酒店</a><a id="1419">银川酒店</a><a id="1420">南昌酒店</a><a id="1421">沈阳酒店</a><a id="1422">石家庄酒店</a><a id="1423">福州酒店</a><a id="1424">合肥酒店</a><a id="1425">南宁酒店</a><a id="1426">乌鲁木齐酒店</a><a id="1427">太原酒店</a><a id="1428">哈尔滨酒店</a><a id="1429">兰州酒店</a><a id="1430">台北酒店</a><a id="1431">拉萨酒店</a><a id="1432">西宁酒店</a><a id="1433">呼和浩特酒店</a><a id="1434">西塘酒店</a><a id="1435">乌镇酒店</a><a id="1436">鼓浪屿酒店</a><a id="1437">朱家尖酒店</a><a id="1438">普陀山酒店</a><a id="1439">千岛湖酒店</a><a id="1440">涠洲岛酒店</a><a id="1441">泸沽湖酒店</a><a id="1442">莫干山酒店</a><a id="1443">南浔酒店</a><a id="1444">北戴河酒店</a><a id="1445">三清山酒店</a><a id="1446">安徽酒店</a><a id="1447">福建酒店</a><a id="1448">甘肃酒店</a><a id="1449">广东酒店</a><a id="1450">广西酒店</a><a id="1451">贵州酒店</a><a id="1452">海南酒店</a><a id="1453">河北酒店</a><a id="1454">黑龙江酒店</a><a id="1455">河南酒店</a><a id="1456">湖北酒店</a><a id="1457">湖南酒店</a><a id="1458">江苏酒店</a><a id="1459">江西酒店</a><a id="1460">吉林酒店</a><a id="1461">辽宁酒店</a><a id="1462">内蒙古酒店</a><a id="1463">宁夏酒店</a><a id="1464">青海酒店</a><a id="1465">山东酒店</a><a id="1466">陕西酒店</a><a id="1467">山西酒店</a><a id="1468">四川酒店</a><a id="1469">台湾酒店</a><a id="1470">新疆酒店</a><a id="1471">西藏酒店</a><a id="1472">云南酒店</a><a id="1473">浙江酒店</a><a id="1474">7天连锁酒店</a><a id="1475">如家快捷酒店</a><a id="1476">汉庭酒店</a><a id="1477">锦江之星</a><a id="1478">香格里拉酒店</a><a id="1479">格林豪泰</a><a id="1480">莫泰</a><a id="1481">布丁酒店</a><a id="1482">桔子酒店</a><a id="1483">希尔顿</a><a id="1484">凯宾斯基</a><a id="1485">皇冠假日酒店</a><a id="1486">假日酒店</a><a id="1487">维也纳酒店</a><a id="1488">星程酒店</a><a id="1489">宜必思酒店</a><a id="1490">洲际酒店</a><a id="1491">万豪酒店</a><a id="1492">速8酒店</a><a id="1493">凯悦酒店</a><a id="1494">百时快捷酒店</a><a id="1495">半岛酒店</a><a id="1496">诺富特酒店</a><a id="1497">城市酒店</a><a id="1498">新加坡酒店</a><a id="1499">迪拜酒店</a><a id="1500">泰国酒店</a><a id="1501">韩国酒店</a><a id="1502">马尔代夫酒店</a><a id="1503">法国酒店</a><a id="1504">锦江之星酒店</a><a id="1505">希尔顿酒店</a><a id="1506">酒店品牌</a></div></div></div><div>特价机票<div><a id="1511">上海机票</a><a id="1512">北京机票</a><a id="1513">广州机票</a><a id="1514">深圳机票</a><a id="1515">重庆机票</a><a id="1516">成都机票</a><a id="1517">昆明机票</a><a id="1518">厦门机票</a><a id="1519">三亚机票</a><a id="1520">杭州机票</a><a id="1521">春秋航空</a><a id="1522">东方航空</a><a id="1523">南方航空</a><a id="1524">中国国航</a><a id="1525">海南航空</a><a id="1526">厦门航空</a><a id="1527">山东航空</a><a id="1528">吉祥航空</a><a id="1529">首都航空</a><a id="1530">打折机票</a><a id="1531">机票地图</a><a id="1532">机场大全</a></div></div><div>旅游度假<div><a id="1537">国内旅游攻略</a><a id="1538">出境旅游攻略</a><a id="1539">马尔代夫旅游</a><a id="1540">普吉岛旅游</a><a id="1541">巴厘岛旅游</a><a id="1542">济州岛旅游</a><a id="1543">日本旅游</a><a id="1544">新加坡旅游</a><a id="1545">澳大利亚旅游</a><a id="1546">迪拜旅游</a><a id="1547">柬埔寨旅游</a><a id="1548">英国旅游</a><a id="1549">法国旅游</a><a id="1550">清迈旅游</a><a id="1551">黄山旅游</a><a id="1552">北京旅游</a><a id="1553">上海旅游</a><a id="1554">杭州旅游</a><a id="1555">苏州旅游</a><a id="1556">成都旅游</a><a id="1557">桂林旅游</a><a id="1558">九寨沟旅游</a><a id="1559">三亚旅游</a><a id="1560">丽江旅游</a><a id="1561">青岛旅游</a><a id="1562">香港旅游</a><a id="1563">澳门旅游</a><a id="1564">台湾旅游</a><a id="1565">时刻表</a></div></div><div>攻略指南<div><a id="1570">攻略社区</a><a id="1571">旅游攻略</a><a id="1572">旅行游记</a><a id="1573">旅游问答</a><a id="1574">周末旅游</a><a id="1575">天气预报</a><a id="1576">地图大全</a><a id="1577">Travel Guide</a><a id="1578">企业商旅</a><a id="1579">中小企业差旅</a><a id="1580">商旅签证</a></div></div><div>快速入口<div><a id="1585">门票导航</a><a id="1586">精选目的地</a><a id="1587">存款证明</a><a id="1588">携程主题曲</a><a id="1589">航班时刻表</a><a id="1590">酒店导航</a><a id="1591">品牌酒店</a><a id="1592">酒店大全</a><a id="1593">酒店经营规则</a><a id="1594">国家酒店</a><a id="1595">eBooking</a><a id="1596">礼品卡</a><a id="1597">企业商旅</a></div></div></div></div><div role="contentinfo" id="1598"><div><div><dl>旅游资讯<a id="1606">宾馆索引</a><a id="1608">攻略索引</a><a id="1610">机票索引</a><a id="1612">网站导航</a><a id="1614">旅游索引</a><a id="1616">邮轮索引</a><a id="1618">企业差旅索引</a></dl><dl>加盟合作<a id="1623">分销联盟</a><a id="1625">友情链接</a><a id="1627">企业礼品卡采购</a><a id="1629">保险代理</a><a id="1631">代理合作</a><a id="1633">酒店加盟</a><a id="1635">目的地及景区合作</a><a id="1637">更多加盟合作</a></dl><dl>关于携程<a id="1642">关于携程</a>携程热点<a id="1648">联系我们</a><a id="1650">诚聘英才</a><a id="1652">用户协议</a><a id="1656">营业执照</a><a id="1658">安全中心</a><a id="1660">携程内容中心</a><a id="1662">知识产权</a><a id="1664">Trip.com Group</a></dl><div><dl><div>境内：<div>95010 或
                                                    400-830-6666</div></div>中国香港：+852-3008-3295 中国澳门：+86-21 3406-4888 中国台湾：+86-21 3406-4888 其他国家和地区：+86-21-3406-4888</dl>扫码下载携程App</div></div><div><p><a id="1694">沪公网备31010502002731号</a>丨<a id="1696">互联网药品信息服务资格证</a>丨<a id="1697">（沪）网械平台备字[2022]第00001号</a>| 沪网食备1050001号 |<a id="1699">旅游度假资质</a>|<a id="1700">平台信息</a></p><p><span>违法和不良信息举报电话021-22500846 丨 全国旅游投诉热线12345 丨</span><a id="1703">上海市旅游网站落实诚信建设主体责任承诺书</a></p><div><a title="工商亮照标识" id="1705">上海工商</a><a title="征信网" id="1706">网络社会征信网</a><a title="上海市互联网违法与违规信息举报中心" id="1707">信息举报中心</a><a id="1708">可信网站</a></div><div><a id="1710">信用中国</a><a id="1711">网上有害信息举报专区</a><a id="1712">消费维权联络点</a><a id="1713">网购大家评</a>适老化及无障碍标识</div></div></div></div></div></div></body>"
`;

exports[`DOMShrinker 应该生成一致的压缩输出并匹配快照 1`] = `
"T1: <span class="author">作者名</span>
T2: <li>要点一</li>
T3: <p>第一段落内容</p>

<body><div class="article"><header><h1>文章标题</h1><div class="meta">{T1(0,1)}<span class="date">2023-04-01</span></div></header><section class="content">{T3(0)}{T3(0)}<ul class="points">{T2(0)}<li id="point-2" class="point">要点二</li>{T2(0)}{T2(0)}</ul></section><footer><div class="tags">{T1(0,1)}{T1(0,1)}{T1(0,1)}</div><button class="like">点赞</button><button class="share">分享</button><div class="link"><a href="https://www.xiaolugu.com">百度</a><a href="https://www.baidu.com">百度</a></div></footer></div></body>"
`;
